crates/assists/src/handlers/reorder_impl.rs

```diff
@@ -0,0 +1,249 @@
+use hir::{PathResolution, Semantics};
+use ide_db::RootDatabase;
+use itertools::Itertools;
+use rustc_hash::FxHashMap;
+use syntax::{
+    algo,
+    ast::{self, NameOwner},
+    AstNode,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Assist: reorder_impl
+//
+// Reorder the methods of a struct impl. If it is an `impl Trait`, then
+// the methods will be ordered in the same order as in the trait definition.
+//
+// ```
+// struct Foo;
+// <|>impl Foo {
+//     fn b() {}
+//     fn c() {}
+//     fn a() {}
+// }
+// ```
+// ->
+// ```
+// struct Foo;
+// impl Foo {
+//     fn a() {}
+//     fn b() {}
+//     fn c() {}
+// }
+// ```
+//
+pub(crate) fn reorder_impl(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let impl_ast = ctx.find_node_at_offset::<ast::Impl>()?;
+    let items = impl_ast.assoc_item_list()?;
+
+    let methods = get_methods(&items);
+    let trait_ = impl_ast.trait_().and_then(|t| match t {
+        ast::Type::PathType(path) => Some(path),
+        _ => None,
+    });
+    let path = trait_.map(|pt| pt.path()).flatten();
+
+    let ranks = compute_method_ranks(&path, ctx, &methods)?;
+
+    let sorted: Vec<_> = methods
+        .iter()
+        .cloned()
+        .sorted_by_key(|f| {
+            f.name().and_then(|n| ranks.get(&n.to_string()).copied()).unwrap_or(usize::max_value())
+        })
+        .collect();
+
+    // Don't edit already sorted methods:
+    if methods == sorted {
+        return None;
+    }
+
+    let target = items.syntax().text_range();
+    acc.add(AssistId("reorder_impl", AssistKind::RefactorRewrite), "Sort methods", target, |edit| {
+        let mut rewriter = algo::SyntaxRewriter::default();
+        for (old, new) in methods.iter().zip(&sorted) {
+            rewriter.replace(old.syntax(), new.syntax());
+        }
+        edit.rewrite(rewriter);
+    })
+}
+
+fn compute_method_ranks(
+    path: &Option<ast::Path>,
+    ctx: &AssistContext,
+    methods: &[ast::Fn],
+) -> Option<FxHashMap<String, usize>> {
+    if let Some(path) = path {
+        let td = trait_definition(path, &ctx.sema)?;
+
+        Some(
+            td.items(ctx.db())
+                .iter()
+                .flat_map(|i| match i {
+                    hir::AssocItem::Function(f) => Some(f),
+                    _ => None,
+                })
+                .enumerate()
+                .map(|(idx, func)| ((func.name(ctx.db()).to_string(), idx)))
+                .collect(),
+        )
+    } else {
+        Some(
+            methods
+                .iter()
+                .flat_map(|f| f.name().map(|n| n.to_string()))
+                .sorted()
+                .enumerate()
+                .map(|(idx, name)| (name, idx))
+                .collect(),
+        )
```

Comment:
Let's maybe remove the case for alphabetic sorting alrogether? 

I feel it's something which you almost never want, but it'll be almost always available.